### PR TITLE
[Segment Replication] Add ClusterState utility to identify SEGMENT replication

### LINKED ISSUE #9566: [Segment Replication] Expose utility to denote whether data can be stale
With https://github.com/opensearch-project/OpenSearch/issues/8211, few plugins resorted to `_primary` preference to support strong reads for all search requests (e.g [sql](https://github.com/opensearch-project/sql/pull/1815)). This is not needed for DOCUMENT replication and thus, puts un-necessary burden on primary. To keep plugins/consumers agnostic of replication strategy, core can expose a utility to identify if replica can return stale data and thus hint consumers to add _primary preference only when needed. 

In core, we can expose a utility which returns boolean denoting that client should use _primary preference if latest data is desired

#### Alternative
1. Do not change anything. All requests will land on primary including requests for DOCUMENT replication type indices
2. Use `_primary` preference only for SEGMENT replication enabled indices. This needs clients to know about underlying replication strategy.

### PR DESCRIPTION
**Pull request by dreamer-89**
### Description
Adds a public method in ClusterState which identifies whether for input index name, `SEGMENT` replication is enabled or not.

### Related Issues
Resolves #9566

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -409,6 +410,15 @@ public boolean supersedes(ClusterState other) {
 
     }
```
Comment (mch2) [inline server/src/main/java/org/opensearch/cluster/ClusterState.java:412]: nit - please add javadoc.

File: server/src/test/java/org/opensearch/cluster/ClusterStateTests.java [hunk 2]
```diff
@@ -84,7 +86,7 @@ public void testSupersedes() {
         final DiscoveryNode node1 = new DiscoveryNode("node1", buildNewFakeTransportAddress(), emptyMap(), emptySet(), version);
         final DiscoveryNode node2 = new DiscoveryNode("node2", buildNewFakeTransportAddress(), emptyMap(), emptySet(), version);
         final DiscoveryNodes nodes = DiscoveryNodes.builder().add(node1).add(node2).build();
-        ClusterName name = ClusterName.CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY);
+        ClusterName name = CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY);
```
Comment (mch2) [inline server/src/test/java/org/opensearch/cluster/ClusterStateTests.java:89]: nit - unrelated here.

**Review by mch2 — APPROVED** [review]
few nits  but otherwise looks good.

File: server/src/test/java/org/opensearch/cluster/ClusterStateTests.java (hunk 2 above)
Comment (dreamer-89) [reply server/src/test/java/org/opensearch/cluster/ClusterStateTests.java:89]: Yup, it is unrelated. This was change automatically, as I introduced new tests which refered CLUSTER_NAME_SETTING

File: server/src/main/java/org/opensearch/cluster/ClusterState.java (hunk 1 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/cluster/ClusterState.java:412]: Added.

**Comment by dreamer-89** [comment]
Previous gradle check failures related to flaky test failures.

Windows precommit failure is due to insufficient memory. 
https://github.com/opensearch-project/OpenSearch/actions/runs/6005796874/job/16289211961?pr=9593
```
> Task :modules:analysis-common:jarHell FAILED
| Output for C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\11.0.20-8\x64\bin\java.exe:OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x0000000090000000, 117440512, 0) failed; error='The paging file is too small for this operation to complete' (DOS error/errno=1455)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 117440512 bytes for Failed to commit area from 0x0000000090000000 to 0x0000000097000000 of length 117440512.
# An error report file with more information is saved as:
# D:\a\OpenSearch\OpenSearch\modules\analysis-common\hs_err_pid2948.log
```

**Review by Rishikesh1159 — APPROVED** [review]

**Pull request merged**